### PR TITLE
Review by 5/13: Feature/6167 regulation search using ecfr api

### LINKED ISSUE #6167: Handle global legal search for regulations
### What we're after:
_eRegulations will be redirecting to the eCFR. Therefore, we will need to update the global legal search to pull results from the eCFR instead of from our own API._

We may want to consider using the eCFR API to return the results for the regulations in our global legal search. Refer to the [eCFR API text search](https://github.com/fecgov/fec-eregs/issues/826#issuecomment-1939134277) research for more information.

Screenshot 2024-03-20 at 8 04 13 PM

### Related issues
- https://github.com/fecgov/fec-eregs/issues/826#issuecomment-1939134277
- https://github.com/fecgov/fec-cms/issues/6161

### Completion criteria
- [x] API call is updated to the new eCFR search 
- [x] Results structure is updated to match eCFR capabilities 

### Tech steps or considerations
- [x] Pull results from eCFR API instead of from out own API 
- [x] Understand the mapping of the results fields between our openFEC API and the new eCFR API - need to display results in a way that makes sense. 
- [x] Handle boolean search operators for regulations. Operators are different for "and" and "or" conditions using the eCFR API.
- [x] Figure out pagination. Pagination is not required in eCFR API.
- [x] There are two different pages that contain results: global legal search (advisory opinions, MURs, statutes, and regulations) and individual regulations search result page (must handle pagination and display of all results).  

### Future work
- [ ] Make improvements to the eRegs search results display based on user feedback and research.
- [ ] Make a design ticket to fix example text, keyword search tooltip, and boolean search pop-up box

### PR DESCRIPTION
**Pull request by patphongs**
## Summary (required)

- Resolves #6167 

Updates: 
- Pull results from eCFR API instead of from our own API 
- Update the mapping of the regulations results to the eCFR API
- Handle boolean search operators for regulations using the view so that the front end operators can remain the same for users.
- Update pagination using eCFR parameters
- Use eCFR API when searching regulations in global legal search (advisory opinions, MURs, statutes, and regulations) and individual regulations search result page.  

### Required reviewers

1 UX
1 Front-end
1 Content

## Impacted areas of the application

General components of the application that this PR will affect:

- [Regulations search](http://127.0.0.1:8000/data/legal/search/regulations/)
- [Global legal search](http://127.0.0.1:8000/legal-resources/)

## Screenshots

There should be no visible changes to the regulations search on our website. The only change is that the results are now coming from the eCFR API.

## How to test

For the content team:
- This has been deployed to feature space. [Global legal search](https://fec-feature-cms.app.cloud.gov/legal-resources/) and [regulations page search](https://fec-feature-cms.app.cloud.gov/data/legal/search/regulations/)

For developers/UX:
- Pull down and run this branch
- Go to try a search on the [global legal search](http://127.0.0.1:8000/legal-resources/) and [regulations page search](http://127.0.0.1:8000/data/legal/search/regulations/)

## System architecture updates (if applicable)

Update system architecture diagram to reflect the eCFR API as one of our external services for regulations search. [Ticket](https://github.com/fecgov/fec-cms/issues/6234).

### CODE REVIEW

**Review by djgarr — APPROVED** [review]
Everything seems to be working properly. Thanks!

**Review by isaacjamesbaker — APPROVED** [review]
Thanks, Pat. I spent a solid 90 minutes searching through the reg links and they all point to eCFR. Things seem to be working well. Thanks so much for your work on this!

File: fec/legal/views.py [hunk 1]
```diff
@@ -274,13 +337,35 @@ def legal_doc_search_af(request):
...
+                "highlights": [obj['full_text_excerpt']],
+                "name": obj['headings']['section'],
+                "no": obj['hierarchy']['section'],
+                "type": None,
+                "url":  (
+                    "https://www.ecfr.gov/current/title-11/"
+                    f"chapter-{obj['hierarchy']['chapter']}/"
+                    f"section-{obj['hierarchy']['section']}"
+                )
+                } for obj in ecfr_results['results']]
```
Comment (johnnyporkchops) [inline fec/legal/views.py:355]: Nice list comprehension!

Comment (johnnyporkchops) [inline fec/legal/views.py:355]: Nice, efficient use of list comprehension!

**Review by johnnyporkchops — COMMENTED** [review]
This is impressive @patphongs! Creating the ecfr-caller based on the api-caller is a great solution to our problem!

File: fec/legal/templates/legal-search-results.jinja [hunk 2]
```diff
@@ -74,7 +74,7 @@
                       {% include 'partials/legal-search-results-regulation.jinja' %}
                       {% endwith %}
                       <div class="results-info">
-                      <a class="button button--browse button--alt" href="/data/legal/search/{{ category }}/?search={{ query }}&search_type={{ category}}">All results</a>
+                      <a class="button button--browse button--alt" href="/data/legal/search/{{ category }}/?search={{ query }}">All results</a>
```
[the remark this answers is not in the document]
Comment (patphongs) [reply fec/legal/templates/legal-search-results.jinja:77]: Thanks for spotting that @johnnyporkchops! I have fixed the all results AO button link

File: fec/legal/views.py [hunk 3]
```diff
@@ -171,7 +231,8 @@ def legal_doc_search_ao(request):
     query = request.GET.get('search', '')
     offset = request.GET.get('offset', 0)
 
-    results = api_caller.load_legal_search_results(query, 'advisory_opinions', offset=offset)
```
[the remark this answers is not in the document]
Comment (patphongs) [reply fec/legal/views.py:174]: Excellent observation @johnnyporkchops, I have put in a follow-up issues to explore how we can also search the labels as well. https://github.com/fecgov/fec-cms/issues/6251

**Comment by patphongs** [comment]
> Have you considered (either for now or for a future enhancement) introducing an icon signifying an external link for the ecfr links? Not sure if this would be necessary or not to address the issue that the experience might be "jarring" for users

@johnnyporkchops I have put in another issue to research this: https://github.com/fecgov/fec-cms/issues/6252

**Review by johnnyporkchops — APPROVED** [review]
Works great, thanks @patphongs

**Pull request merged**